### ISSUE TITLE
feat: line specific rule toggling

### ISSUE THREAD

**Issue by FalkF** [rwjblue/ember-template-lint#354]
Like other linters, ember-template-lint should allow users to enable and disable rules per line:

```{{! template-lint-enable-line bare-strings }}```

```{{! template-lint-disable-line bare-strings }}```

```{{! template-lint-enable-next-line bare-strings }}```

```{{! template-lint-disable-next-line bare-strings }}```

**Comment by rwjblue** [comment]
It is already possible to enable / disable via comments. Are you specifically saying that you want the addition of a single line comment?

**Comment by FalkF** [comment]
Yes, disabling per line would be nice.
This way one can make exceptions to a rule without disabling it for a whole file.
Maybe I manage to implement this :)

**Comment by rwjblue** [comment]
You can already make exceptions to a single rule without disabling it for a whole file. 🤔

**Comment by rwjblue** [comment]
Note: I’m not opposed to the capability, I’m just not sure how to make it work properly. “Next line” is pretty ambiguous when talking about ASTs...

**Comment by rwjblue** [comment]
I'm still curious about what you have in mind here.

@FalkF - Are you still interested in this one?

**Comment by Redsandro** [comment]
> You can already make exceptions to a single rule without disabling it for a whole file. :thinking:

How?

**Comment by rwjblue** [comment]
It is documented in [this section of the README](https://github.com/ember-template-lint/ember-template-lint#per-template) though maybe we need to make it even clearer. Would definitely like help.


---

An instruction will apply to all later siblings and their descendants:

```hbs
<!-- disable for <p> and <span> and their contents, but not for <div> or <hr> -->
<div>
  <hr>
  {{! template-lint-disable }}
  <p>
    <span>Hello!</span>
  </p>
</div>
```

An in-element instruction will apply to only that element:

```hbs
<!-- enable for <p>, but not for <div>, <hr> or <span> -->
<div>
  <hr>
  <p {{! template-lint-enable }}>
    <span>Hello!</span>
  </p>
</div>
```

An in-element instruction with the `-tree` suffix will apply to that element and all its descendants:

```hbs
<!-- configure for <p>, <span> and their contents, but not for <div> or <hr> -->
<div>
  <hr>
  <p {{! template-lint-configure-tree block-indentation "tab" }}>
    <span>Hello!</span>
  </p>
</div>
```

**Comment by Redsandro** [comment]
@rwjblue Hmm, I must have missed that. Apologies. 

> maybe we need to make it even clearer. Would definitely like help.

So I skimmed though the readme, skipped the part about configuration, went straight for the rules, clicked a bunch of links, and did not find what I was looking for.

The section about disabling rules should be a subparagraph for the rules section. People don't read an entire doc; they skim the paragraphs that are most likely to contain what they are looking for.

PR: #576

**Comment by patocallaghan** [comment]
@rwjblue we've came across an instance where having the `template-lint-disable-next-line`, similar to `stylelint-disable-next-line` and `eslint-disable-next-line`, would be extremely useful for us. 

Currently where I work we follow a process of when we encounter a best practice, rather than having to refactor all our code at once we'll enable a linting rule to prevent it, but disable it inline for existing usages, meaning new additions will be caught out by the lint rule.

Using `{{! template-lint-disable some-rule }}` and its variations rely on us having to disable and re-enable lots of rules to make sure we only target the specific existing instances. On the other hand having a `-next-line` variation would allow us to be more specific about what we are disabling and also reduce our required comment statements by half as we no longer need to re-enable them.

Take the following examples, we don't want other instances of these violations slipping through so we try to be very pointed with our disabling statements.

### What we currently do
```hbs
{{! template-rule-disable my-plugin/no-deprecated-classname }}
{{#some-component class="some-deprecated-classname"}}
{{! template-rule-enable my-plugin/no-deprecated-classname }}
  <div>...</div>
  {{! template-rule-disable my-plugin/no-deprecated-classname }}
  <div class="some-deprecated-classname">
  {{! template-rule-enable my-plugin/no-deprecated-classname }}
  </div>
{{/some-component}}

{{! template-rule-disable my-plugin/no-deprecated-classname }}
<div class="some-deprecated-classname">
{{! template-rule-enable my-plugin/no-deprecated-classname }}
...
</div>
```

### What we'd prefer
```hbs
{{! template-rule-disable-next-line my-plugin/no-deprecated-classname }}
{{#some-component class="some-deprecated-classname"}}
  <div>...</div>
  {{! template-rule-disable-next-line my-plugin/no-deprecated-classname }}
  <div class="some-deprecated-classname">
  </div>
{{/some-component}}

{{! template-rule-disable-next-line my-plugin/no-deprecated-classname }}
<div class="some-deprecated-classname">
...
</div>
```

You mention

> “Next line” is pretty ambiguous when talking about ASTs

but can we not use the same mechanism that ESLint and Stylelint, i.e. the `loc` object?

I'd be happy to work on this if you think it were a worthy addition and somewhat doable.

**Comment by rwjblue** [comment]
> Using {{! template-lint-disable some-rule }} and its variations rely on us having to disable and re-enable lots of rules to make sure we only target the specific existing instances.

Hmm, this isn't necessarily true. The rule enabling / disabling is based on natural scoping. For example, this example from the README:

```hbs
<!-- disable for <p> and <span> and their contents, but not for <div> or <hr> -->
<div>
  <hr>
  {{! template-lint-disable }}
  <p>
    <span>Hello!</span>
  </p>
</div>
```

**Comment by patocallaghan** [comment]
Yep that's true, but that's not the behaviour I was hoping for. Please correct me if this is wrong 😅 

If we take the following contrived example, we have a custom lint to prevent the introduction of `some-deprecated-classname`. We've opted out the existing violation with `{{! template-lint-disable my-plugin/no-deprecated-classname}}`

```hbs
<div>
  <hr>
  {{! template-lint-disable my-plugin/no-deprecated-classname}}
  <p class="some-deprecated-classname">
    <span>Hello!</span>
  </p>
</div>
```

But then the next day someone else comes along and adds a new instance of it like so:

```hbs
<div>
  <hr>
  {{! template-lint-disable my-plugin/no-deprecated-classname}}
  <p class="some-deprecated-classname">
    <span>Hello!</span>
    <div class="some-deprecated-classname"></div> // We want this to error
  </p>
</div>
```

We want the lint rule to fire for the new instance but it doesn't because it's disabled the rule for the `<p>` and all of its contents. To get it to fire we've had to do.

```hbs
<div>
  <hr>
  {{! template-lint-disable my-plugin/no-deprecated-classname}}
  <p class="some-deprecated-classname">
  {{! template-lint-enable my-plugin/no-deprecated-classname}}
    <span>Hello!</span>
    <div class="some-deprecated-classname"></div>
  </p>
</div>
```

While this is a simple example our templates are more complex and deeply nested. Having something like `template-lint-disable-next-line` I'd imagine would enable us to do this easier.

**Comment by rwjblue** [comment]
IMHO, we should be using the `pending` stuff (in `.template-lintrc.js`) for this kind of "gradual lockdown" type thing. The rough selling points are:

* All new templates are authored with the most strict settings.
* Existing templates can't "get worse" by regressing additonal rules (this is why its not "completely ignore").
* The "pending" list gives you a _perfect_ TODO list.
* When a file begins passing the lint config (e.g. it is no longer "pending") it will start failing (with the message of something like "please remove this file from the pending list").
* Doesn't clutter up your individual templates with loads of enable/disable config.

I think the only thing we'd need to do in addition to what already exists is the ability to track specific lines for each rule violation (currently you can just disable the rule for the full file).

**Comment by patocallaghan** [comment]
Oh interesting, I never knew about `pending`. So I guess one downside to the pending/line number combo is if the markup is refactored or moved around rules, would start failing.

We're used to the `disable-next-line` pattern in our .scss and .js files already, but if you are completely against it adding it I can look at the `pending` stuff.

**Comment by patocallaghan** [comment]
It looks like `pending` emits warnings for each of the violations listed. Unfortunately for our use case this is not a great experience considering we have thousands of these. It also looks like using `pending` they still show up as linting errors in vscode (possibly a bug in ember language server).

![image](https://user-images.githubusercontent.com/685034/58992460-cc369a80-87e2-11e9-9430-3a06e1927698.png)

**Comment by 22a** [comment]
@FalkF @rwjblue curiosity got the best of me and I've spiked out what this could potentially look like in https://github.com/ember-template-lint/ember-template-lint/pull/780 let us know what you think, if this is something you'd like to see in ember-template-lint

**Comment by rafaelbnp** [comment]
> It is documented in [this section of the README](https://github.com/ember-template-lint/ember-template-lint#per-template) though maybe we need to make it even clearer. Would definitely like help.

is this still documented somewhere? I couldn't find anywhere in the README 🤔

**Comment by bertdeblock** [comment]
@rafaelbnp Should be here => https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/configuration.md#per-template-file.
